### PR TITLE
Refactor Initialization of Configuration File Merger

### PR DESCRIPTION
**Pull request by dnestoro**
Remove macro from merger init and throw better error if executable does not exist

### CODE REVIEW

File: native-maven-plugin/src/main/java/org/graalvm/buildtools/maven/config/AbstractMergeAgentFilesMojo.java [hunk 1]
```diff
@@ -58,32 +58,24 @@ public abstract class AbstractMergeAgentFilesMojo extends AbstractMojo {
...
         }
 
-        File nativeImageExecutable = nativeImageExecutablePath.toAbsolutePath().toFile();
+        return mergerExecutable;
+    }
+
+    private void initializeMergerExecutable() throws MojoExecutionException {
+        Path nativeImage = NativeImageConfigurationUtils.getNativeImage(logger);
+        File nativeImageExecutable = nativeImage.toAbsolutePath().toFile();
         File mergerExecutable = new File(nativeImageExecutable.getParentFile(), nativeImageConfigureFileName());
```
Comment (fniephaus) [inline native-maven-plugin/src/main/java/org/graalvm/buildtools/maven/config/AbstractMergeAgentFilesMojo.java:74]: Use a variable for `nativeImageConfigureFileName()` so it's only called once.

File: native-maven-plugin/src/main/java/org/graalvm/buildtools/maven/config/AbstractMergeAgentFilesMojo.java [hunk 2]
```diff
@@ -58,32 +58,24 @@ public abstract class AbstractMergeAgentFilesMojo extends AbstractMojo {
...
-                if (installProcess.waitFor() != 0) {
-                    getLog().warn("Installation of native image merging tool failed");
-                }
-                NativeImageUtils.maybeCreateConfigureUtilSymlink(mergerExecutable, nativeImageExecutablePath);
-            } catch (IOException | InterruptedException e) {
-                // ignore since we will handle that if the installer doesn't exist later
-            }
-
+            throw new MojoExecutionException("'" + nativeImageConfigureFileName() + "' tool was not found in your " + nativeImage + "." +
+                    "This probably means that the JDK at '" + nativeImage + "' is not a GraalVM distribution."
```
Comment (fniephaus) [inline native-maven-plugin/src/main/java/org/graalvm/buildtools/maven/config/AbstractMergeAgentFilesMojo.java:77]: `getNativeImage()` already checks whether `native-image` exists, so if `native-image-configure` is missing, then it is an incomplete GraalVM distribution. Maybe something like this:

```suggestion
            throw new MojoExecutionException("'" + nativeImageConfigureFileName() + "' tool was not found in the GraalVM JDK at '" + nativeImageExecutable.getParentFile().getParentFile() + "'." +
                    "This probably means that you are using a GraalVM distribution that is not fully supported by the Native Build Tools."
```

**Comment by dnestoro** [comment]
@fniephaus @melix @alvarosanchez please take a look on this PR when you catch some time. I would like to merge it if it looks okay (since it is open while ago)

File: native-maven-plugin/src/main/java/org/graalvm/buildtools/maven/config/AbstractMergeAgentFilesMojo.java [hunk 3]
```diff
@@ -58,32 +58,25 @@ public abstract class AbstractMergeAgentFilesMojo extends AbstractMojo {
...
+    }
 
+    private void initializeMergerExecutable() throws MojoExecutionException {
+        Path nativeImage = NativeImageConfigurationUtils.getNativeImage(logger);
+        File nativeImageExecutable = nativeImage.toAbsolutePath().toFile();
+        String nativeImageConfigureFileName = nativeImageConfigureFileName();
+        File mergerExecutable = new File(nativeImageExecutable.getParentFile(), nativeImageConfigureFileName);
+        if (!mergerExecutable.exists()) {
+            throw new MojoExecutionException("'" + nativeImageConfigureFileName + "' tool was not found in the GraalVM JDK at '" + nativeImageExecutable.getParentFile().getParentFile() + "'." +
+                    "This probably means that you are using a GraalVM distribution that is not fully supported by the Native Build Tools."
```
Comment (fniephaus) [inline native-maven-plugin/src/main/java/org/graalvm/buildtools/maven/config/AbstractMergeAgentFilesMojo.java:78]: Can we make this actionable, for example like this:
```suggestion
                    "This probably means that you are using a GraalVM distribution that is not fully supported by the Native Build Tools. Please try again, for example, with Oracle GraalVM or GraalVM Community Edition."
```

**Review by alvarosanchez — DISMISSED** [review]

File: native-maven-plugin/src/main/java/org/graalvm/buildtools/maven/config/AbstractMergeAgentFilesMojo.java [hunk 4]
```diff
@@ -58,32 +58,25 @@ public abstract class AbstractMergeAgentFilesMojo extends AbstractMojo {
...
-                }
-                NativeImageUtils.maybeCreateConfigureUtilSymlink(mergerExecutable, nativeImageExecutablePath);
-            } catch (IOException | InterruptedException e) {
-                // ignore since we will handle that if the installer doesn't exist later
-            }
+        return mergerExecutable;
+    }
 
+    private void initializeMergerExecutable() throws MojoExecutionException {
+        Path nativeImage = NativeImageConfigurationUtils.getNativeImage(logger);
```
Comment (fniephaus) [inline native-maven-plugin/src/main/java/org/graalvm/buildtools/maven/config/AbstractMergeAgentFilesMojo.java:72]: Could you also improve the error in `getNativeImage()` please? Something like:
```java
            throw new RuntimeException("The 'native-image' tool was not found on your system. " + System.lineSeparator() +
                    "Make sure that the JAVA_HOME or GRAALVM_HOME environment variables point to a GraalVM JDK, or that 'native-image' is on the system path.");
```

File: native-maven-plugin/src/main/java/org/graalvm/buildtools/maven/config/AbstractMergeAgentFilesMojo.java [hunk 5]
```diff
@@ -58,32 +58,25 @@ public abstract class AbstractMergeAgentFilesMojo extends AbstractMojo {
...
+        File nativeImageExecutable = nativeImage.toAbsolutePath().toFile();
+        String nativeImageConfigureFileName = nativeImageConfigureFileName();
+        File mergerExecutable = new File(nativeImageExecutable.getParentFile(), nativeImageConfigureFileName);
+        if (!mergerExecutable.exists()) {
+            throw new MojoExecutionException("'" + nativeImageConfigureFileName + "' tool was not found in the GraalVM JDK at '" + nativeImageExecutable.getParentFile().getParentFile() + "'." +
```
Comment (fniephaus) [inline native-maven-plugin/src/main/java/org/graalvm/buildtools/maven/config/AbstractMergeAgentFilesMojo.java:77]:
```suggestion
            throw new MojoExecutionException("The '" + nativeImageConfigureFileName + "' tool was not found in the GraalVM JDK at '" + nativeImageExecutable.getParentFile().getParentFile() + "'." +
```

**Review by melix — DISMISSED** [review]

File: native-maven-plugin/src/main/java/org/graalvm/buildtools/utils/NativeImageConfigurationUtils.java [hunk 6]
```diff
@@ -128,8 +128,8 @@ public static Path getNativeImage(Logger logger) throws MojoExecutionException {
         }
 
         if (nativeImage == null) {
-            throw new RuntimeException("GraalVM native-image is missing on your system. " + System.lineSeparator() +
-                    "Make sure that GRAALVM_HOME environment variable is present.");
+            throw new RuntimeException("The 'native-image' tool was not found on your system. " + System.lineSeparator() +
```
Comment (fniephaus) [inline native-maven-plugin/src/main/java/org/graalvm/buildtools/utils/NativeImageConfigurationUtils.java:131]: Minor consistency nit, maybe remove the newline:
```suggestion
            throw new RuntimeException("The 'native-image' tool was not found on your system. " +
```

**Review by fniephaus — APPROVED** [review]
LGTM, thanks!

**Review by fniephaus — APPROVED** [review]

**Review by fniephaus — APPROVED** [review]

**Pull request merged**